Arm/Insts/DPI/Bitfield.lean

```diff
@@ -0,0 +1,83 @@
+/-
+Copyright (c) 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Shilpi Goel, Nevine
+-/
+-- For now, support only UBFM (immediate) aliased as LSR (immediate):
+-- 32- and 64-bit versions
+
+import Arm.Decode
+import Arm.Insts.Common
+
+namespace DPI
+
+open Std.BitVec
+
+@[simp]
+def exec_bitfield (inst: Bitfield_cls) (s : ArmState) : ArmState :=
+  -- TODO: match opc
+  let immr5 := inst.immr >>> 5
+  let imms5 := inst.imms >>> 5
+  if (inst.sf == 1 && inst.N != 1) ||
+     (inst.sf == 0 && (inst.N != 0 || immr5 != 0 || imms5 != 0)) then
+    write_err (StateError.Illegal s!"Illegal {inst} encountered!") s
+  else
+    let datasize  := if inst.sf == 1#1 then 64 else 32
+    let wtmask    := decode_bit_masks inst.N inst.imms inst.immr false datasize
+    match wtmask with
+    | none => write_err (StateError.Illegal s!"Illegal {inst} encountered!") s
+    | some (wmask, tmask) =>
+      let src := read_gpr datasize inst.Rn s
+      let bot := (BitVec.ror src inst.immr.toNat) &&& wmask
+      let result := bot &&& tmask
+      let s := write_gpr datasize inst.Rd result s
+      let s := write_pc ((read_pc s) + 4#64) s
+      s
+
+
+----------------------------------------------------------------------
+
+/-- Generate random instructions of the DPI.Bitfield class. -/
+partial def bitfield_rand_aux (sf : BitVec 1) (N : BitVec 1) (immr : BitVec 6) (imms : BitVec 6) :
+IO (Option (BitVec 32)) := do
+  if sf == 1 && N != 1 then
+    let N := 0b1#1
+    bitfield_rand_aux sf N immr imms
+  else if sf == 0 && (N != 0 || extractLsb 5 5 immr != 0 || extractLsb 5 5 imms != 0) then
+    let sf := ← BitVec.rand 1
+    let N := sf
+    let immr := sf ++ (← BitVec.rand 5)
+    let imms := sf ++ (← BitVec.rand 5)
+    bitfield_rand_aux sf N immr imms
+  else
+  let (inst : Bitfield_cls) :=
+    { sf    := sf,
+      opc   := ← pure 0b10#2,
+      N     := N,
+      immr  := immr,
+      imms  := imms,
+       -- TODO: Do we need to limit Rn and Rd to be up to 30 as in Add_sub_imm?
+      Rn    := ← BitVec.rand 5,
+      Rd    := ← BitVec.rand 5 }
+  pure (some (inst.toBitVec32))
+
+partial def Bitfield_cls.ubfm.rand : IO (Option (BitVec 32)) := do
+  let sf := ← BitVec.rand 1
+  let N := ← BitVec.rand 1
+  let immr := ← BitVec.rand 6
+  let imms := ← BitVec.rand 6
+  bitfield_rand_aux sf N immr imms
+
+partial def Bitfield_cls.lsr.rand : IO (Option (BitVec 32)) := do
```

Comment:
How about defining a single function called `Bitfield_cls.ubfm.rand` that tests just UBFM? Since it has a bunch of alias'es, it will cover the testing of all of them (LSR variants, LSL variants, UBFIZ, UBFX, UXTB and UXTH), including LSR.

Reply:
That is true, I just wanted to ensure that we're testing LSR since it's the instruction we're targeting. The conditions for it are not that likely to occur imms = sf|11111, so I wanted to have a function to enforce it. However, after I directed the second trials in the else if branch in bitfield_rand_aux, I started seeing more LSR and LSL in the conformance test output. Still with the latest change, I don't see that many, but I think some are assembled as UBFIZ and UBFX.

Reply:
Thanks for the explanation. I was wondering what you meant in the PR description. I think it is a valid concern. The problem is that the logic in `bitfield_rand_aux` is a bit too complex. Do you think if it would be better to have two separate functions (instead of using `bitfield_rand_aux`), one for `ubfm` and one for `lsr`, just to make the logic simpler? And add a comment for the `lsr` function that explains why it is treated special.

Reply:
I thought to do that at the beginning, but then did it this way to avoid repeated checks and code, but I can go back to separating them as I think now that the checks for invalid instructions will be only in the `ubfm` function. The only repeated code will be to create `inst`. Is it preferable to separate them?

Reply:
I would say yes. @shigoel ?

Reply:
Done. Thank you, Yan, for the advice. It is more readable now and it made me notice that the LSR generation wasn't purely for LSR because of `immr` being generates as 6 random bits instead of making the most significant match `sf`. Now it is only generating LSR instructions as desired.

Reply:
Thanks, @nebeid and @pennyannn, and sorry for being late on this!

The `UBFM` and `LSR` random generation functions look very clear now, thanks. However, this does open us up to defining separate generation functions for each instruction alias. Sure, we only care about `LSR` right now, but we may very well want `LSL`, `UXTH`, etc. to be tested too in the future. That solution makes me uncomfortable.

If we increase the number of random tests, we won't need to do the above. Also, we don't have conformance testing running in CI yet, but we will at some point this year --- the issue of not hitting certain values randomly won't be as important because of the volume of tests we'd be running.

For now, we could leave this as is, but I will revisit this later.